### DISCUSSION TITLE
Understanding the Interaction Between the mint Function's onlyOwner Modifier in DecentralizedStableCoin and Its Use in DSCEngine

### DISCUSSION THREAD

**Discussion by prjp960** [Cyfrin/foundry-full-course-cu#1620] [Q&A] [answered]
how is it possible for anyone to call the mintDsc function and mint tokens even though the mint function in DecentralizedStableCoin is protected by onlyOwner
 the code from DecentralisedStableCoin.sol
```solidity
 function mint(address _to, uint256 _amount) external onlyOwner returns (bool) {
        if (_to == address(0)) {
            revert DecentralizedStableCoin__NotZeroAddress();
        }
        if (_amount <= 0) {
            revert DecentralizedStableCoin__AmountMustBeMoreThanZero();
        }
        _mint(_to, _amount);
        return true;
    }
```
    and this is DSCEngine.sol
```solidity
 DecentralizedStableCoin private immutable i_dsc;
 constructor(address[] memory tokenAddresses, address[] memory priceFeedAddresses, address dscAddress) {
        if (tokenAddresses.length != priceFeedAddresses.length) {
            revert DSCEngine__TokenAddressesAndPriceFeedAddressesAmountsDontMatch();
        }
        // These feeds will be the USD pairs
        // For example ETH / USD or MKR / USD
        for (uint256 i = 0; i < tokenAddresses.length; i++) {
            s_priceFeeds[tokenAddresses[i]] = priceFeedAddresses[i];
            s_collateralTokens.push(tokenAddresses[i]);
        }
        i_dsc = DecentralizedStableCoin(dscAddress);
    }
function mintDsc(uint256 amountDscToMint) public moreThanZero(amountDscToMint) nonReentrant {
        s_DSCMinted[msg.sender] += amountDscToMint;
        revertIfHealthFactorIsBroken(msg.sender);
        bool minted = i_dsc.mint(msg.sender, amountDscToMint);

        if (minted != true) {
            revert DSCEngine__MintFailed();
        }
    }
```

**Comment by 0fprod** [comment, marked as answer]
Because the owner of the `DecentralisedStableCoin` must be the `DSCEngine`. I recommend you to write unit tests about this to fully understand whats happening :).

The DecentralisedStableCoin has a method called `transferOwnership(address)`. You can use this to set the Engine as the owner.
This is how I did it if you wanna take a look https://github.com/0fprod/defi-protocol/blob/main/test/DSCEngine.t.sol#L51C9-L51C48

**Reply by prjp960** [reply]
Thank you now i noticed the transferOwnerShip so after you've deployed those 2 contracts you must've to transfer the ownership to DSCEngine in order to have an access for ownable functions. and i'm really appreciate your answer

**Comment by EngrPips** [comment]
Hello @prjp960, @0fprod  gave a good answer, but if there is still anything that needs to be clarified, please feel free to ask.
